service/history/handler.go

```diff
@@ -1144,12 +1144,36 @@ func (h *handlerImpl) SignalWithStartWorkflowExecution(
 		return nil, h.error(err1, scope, domainID, workflowID)
 	}
 
-	resp, err2 := engine.SignalWithStartWorkflowExecution(ctx, wrappedRequest)
-	if err2 != nil {
-		return nil, h.error(err2, scope, domainID, workflowID)
-	}
+	for {
+		resp, err2 := engine.SignalWithStartWorkflowExecution(ctx, wrappedRequest)
+		if err2 == nil {
+			return resp, nil
+		}
+		// Two simultaneous SignalWithStart requests might try to start a workflow at the same time.
+		// This can result in one of the requests failing with one of two possible errors:
+		//    1) If it is a brand new WF ID, one of the requests can fail with WorkflowExecutionAlreadyStartedError
+		//       (createMode is persistence.CreateWorkflowModeBrandNew)
+		//    2) If it an already existing WF ID, one of the requests can fail with a CurrentWorkflowConditionFailedError
+		//       (createMode is persisetence.CreateWorkflowModeWorkflowIDReuse)
+		// If either error occurs, just go ahead and retry. It should succeed on the subsequent attempt.
+		// For simplicity, we keep trying unless the context finishes or we get an error that is not one of the
+		// two mentioned above.
+		_, isExecutionAlreadyStartedErr := err2.(*persistence.WorkflowExecutionAlreadyStartedError)
```

Comment:
nit: prefer `errors.Is` I think, that way it'll keep working even if we end up wrappign the error

Reply:
will do